File: PRACTICAS/practica3/Civitas_Ruby/pkg/Civitas_Ruby-0.0.1/lib/jugador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'diario'
require_relative 'titulo_propiedad.rb'

module Civitas

  class Jugador
    
    @@Casas_max = 4
    @@Casas_por_hotel = 4
    @@Hoteles_max = 4
    @@Paso_por_salida = 100
    @@Precio_libertad = 200
    @@Saldo_inicial = 7500
    
    public    
    def <=>(jugador)
     
      if @saldo.to_i > jugador.saldo.to_i
        return -1
      elsif @saldo.to_i == jugador.saldo.to_i
        return 0
      end
      
      return 1
            
    end
    
    public
    def en_bancarrota
      
      return @saldo <= 0  
      
    end
    public
    def mover_a_casilla(num_casilla)
      
      if !@encarcelado
        @num_casilla_actual = num_casilla
        @puede_comprar = false 
        Diario.instance.ocurre_evento("El jugador #{@nombre} ha sido movido a la casilla #{num_casilla}");
        return true
      else
        Diario.instance.ocurre_evento("El jugador #{@nombre} no ha sido movido a la casilla #{num_casilla}");
        return false
      end    
      
    end
    def initialize(encarcelado = false , nombre = "Undefined", num_casilla_actual = 0, puede_comprar = true, saldo = @@Saldo_inicial, propiedades = Array.new, salvoconducto = nil)
          
        @encarcelado = encarcelado
        @nombre = nombre
        @num_casilla_actual = num_casilla_actual
        @puede_comprar = puede_comprar
        @saldo = saldo
        @propiedades = propiedades
        @salvoconducto = salvoconducto   
        
    end
    public
    attr_reader :encarcelado, :num_casilla_actual, :puede_comprar, :salvoconducto, :Casas_por_hotel, :nombre, :propiedades, :saldo
 
    
    private
    attr_reader :Casas_max, :Hoteles_max, :Precio_libertad, :Paso_por_salida
    private :Casas_max, :Hoteles_max, :Precio_libertad, :Paso_por_salida
    
    public
    
    def self.jugador1(nombre)
      self.new(false, nombre, 0, true, @@Saldo_inicial, Array.new, nil)
    end
    protected
    def self.jugador2(jugador)    
      self.new(jugador.encarcelado, jugador.nombre, jugador.num_casilla_actual, jugador.puede_comprar, jugador.saldo, jugador.propiedades, jugador.salvoconducto)   
    end
      
    def debe_ser_encarcelado
        
      if !@encarcelado
        if @salvoconducto == nil    
          return true  
        else
        perder_salvoconducto
        Diario.instance.ocurre_evento("El jugador #{@nombre} se libra de la carcel y pierde salvoconducto.")
        return false  
        end
      else
        return false
      end
      
    end
    
    public
    
    def encarcelar(num_casilla_carcel)
      if debe_ser_encarcelado        
        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true
        Diario.instance.ocurre_evento("El jugador #{@nombre} ha sido trasladado a la carcel.")   
        else
          Diario.instance.ocurre_evento("El jugador #{@nombre} no ha sido trasladado a la carcel.")
        end
      end
      return @encarcelado
    end
    
    def obtener_salvoconducto(sorpresa)
      
      if !@encarcelado
        @salvoconducto = sorpresa
        return true
      else
        return false
      end
   
    end
    
    private
    
    def perder_salvoconducto
      
      @salvoconducto.usada
      @salvoconducto = nil
      
    end
    
    public
    
    def tiene_salvoconducto
      
      return @salvoconducto != nil
      
    end
    
    def puede_comprar_casilla
      @puede_comprar = !@encarcelado
    
      return @puede_comprar
      
    end
    
    def paga(cantidad)
      
      return modificar_saldo(cantidad*-1)
      
    end
    
    def paga_impuesto(cantidad)
      if @encarcelado
        return false
      else
        return paga(cantidad)
      end 
    end
    
    def paga_alquiler(cantidad)
      if @encarcelado
        return false
      else
        return paga(cantidad)
      end
    end
    
    def recibe(cantidad)
      if @encarcelado
        return false
      else
        return modificar_saldo(cantidad)
      end
    end
      
    def modificar_saldo(cantidad)
      @saldo += cantidad
      Diario.instance.ocurre_evento("El saldo del jugador #{@nombre} ha sido modificado a #{@saldo}")
      return true
    end
    
    
    private
    
    def puedo_gastar(precio)     
      if !@encarcelado
            return @saldo >= precio;
      else
            return false;
      end
    end
    
    public
    
    def vender(ip)
      
      if !@encarcelado && existe_la_propiedad(ip)
        vender = @propiedades.at(ip).vender(self)
        
        if vender
          Diario.instance.ocurre_evento("El jugador #{@nombre} ha vendido la propiedad #{@propiedades.at(ip).nombre}")
          @propiedades.delete_at(ip)
          return true
        end
      end
      Diario.instance.ocurreEvento("El jugador #{@nombre} no ha podido vender la propiedad (porque estaba encarcelada o no es suya) #{@propiedades.at(ip).nombre}")      
      return false
        
    end
    
    def tiene_algo_que_gestionar
      
      return @propiedades != nil
      
    end
    
    private
    
    def puede_salir_carcel_pagando      
      return @saldo >= @@Precio_libertad   
    end
    
    public
    
    def salir_carcel_pagando
          
      if @encarcelado and puede_salir_carcel_pagando
        paga(@@Precio_libertad)
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugador #{@nombre} sale de la carcel pagando")
        return true
      else
        return false
      end
      
    end
    
    def salir_carcel_tirando

      if Dado.instance.salgo_de_la_carcel
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugador #{@nombre} sale de la carcel tirando")
        return true
      else
        return false
        
      end
      
    end
    
    def pasa_por_salida

      modificar_saldo(@@Paso_por_salida)
      Diario.instance.ocurre_evento("El jugador #{@nombre} pasa por salida y cobra 100")
      return true 
      
    end
    
    def cantidad_casas_hoteles
      
      suma = 0
      
      for i in (0..@propiedades.size-1)
        suma += @propiedades.at(i).cantidad_casas_hoteles
      end
      
      return suma
      
    end
    
    
    
    private
    
    def existe_la_propiedad(ip)
       return ip >= 0 && ip < @propiedades.size
    end
    
    public
    
    def is_encarcelado
      return @encarcelado
    end
    
    private
    
    def puedo_edificar_casa(propiedad) #eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee

      precio = propiedad.precio_edificar

      return propiedad.es_este_el_propietario(self) && propiedad.num_casas < @@Casas_por_hotel && self.puedo_gastar(precio)

    end

    def puedo_edificar_hotel(propiedad)

        precio = propiedad.precio_edificar

      return propiedad.num_casas >=  @@Casas_por_hotel && propiedad.num_hoteles < @@Casas_por_hotel && self.puedo_gastar(precio)

    end
    
    

    def to_s 
         
      return "Jugador de nombre #{@nombre} situado en la casilla #{@num_casilla_actual} con saldo #{@saldo}"   
    
    end
   
    
    def construir_casa(ip)
        result =false 
        puedo_edificar_casa =false
        
        if(@encarcelado)
            return result
        else
            
            existe = existe_la_propiedad(ip)
        
            if(existe)
                
                propiedad = @propiedades.at(ip)
                puedo_edificar_casa=puedo_edificar_casa(propiedad)
                precio = propiedad.precio_edificar
                if(puedo_gastar(precio) && propiedad.num_casas < @@Casas_max)
                    
                    puedo_edificar_casa=true
                    
                end
                if(puedo_edificar_casa)
                    
                    result=propiedad.construir_casa(self)
                
                    if(result)
                        
                        Diario.instance.ocurre_evento("El jugador #{@nombre} construye casa en la propiedad #{@propiedades.at(ip)}")
                        
                    end
                end
            end
        end
        return result
    end
    
    def construir_hotel(ip)
        result = false
        
        if(@encarcelado)
            return result
        end
        
        if existe_la_propiedad(ip)
            
          propiedad = @propiedades.at(ip)
            
          puedo_edificar_hotel = puedo_edificar_hotel(propiedad)
   
          if(puedo_edificar_hotel)
                
          result = propiedad.construir_hotel(self)
          propiedad.derruir_casas(@@Casas_por_hotel, self)
                
          end
          
          Diario.instance.ocurre_evento("El jugador #{@nombre} costruye hotel en la propiedad #{@propiedades.at(ip)}")
          
          end
        
          return result
            
    end
    
    def hipotecar(ip)
        
        result=false
        if @encarcelado
            
            return result
            
        end
        
        if existe_la_propiedad(ip)
            
            propiedad = @propiedades.at(ip)
            result=propiedad.hipotecar(self)
                     
        end
        if result
            
            Diario.instance.ocurre_evento("El jugador #{@nombre} hipoteca la propiedad #{@propiedades.at(ip).nombre}") 
        else
            Diario.instance.ocurre_evento("El jugador #{@nombre} no puede hipotecar la propiedad #{@propiedades.at(ip).nombre}") 
        end
        
        return result
    end
      
    def cancelar_hipoteca(ip)
    
        result=false
        
        if @encarcelado
            
            return result
            
        end
        if(existe_la_propiedad(ip))
            
            propiedad = @propiedades.at(ip)
            cantidad = propiedad.get_importe_cancelar_hipoteca
            puedo_gastar=puedo_gastar(cantidad)
            if puedo_gastar
                
                result = propiedad.cancelar_hipoteca(self)
                                
            
              if result
                Diario.instance.ocurre_evento("El jugador #{@nombre} deshipoteca la propiedad #{@propiedades.at(ip).nombre}") 
              else
                Diario.instance.ocurre_evento("El jugador #{@nombre} no puede cancelar la hipoteca la propiedad #{@propiedades.at(ip).nombre}") 
              end
        else
          Diario.instance.ocurre_evento("El jugador #{@nombre} no puede cancelar la hipoteca la propiedad #{@propiedades.at(ip).nombre}") 
        end
    
        end
    
    return result
  end
    
    
end

File: PRACTICAS/practica3/Civitas_P1/src/civitas_p1/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas_p1;

import static java.lang.Float.compare;
import java.util.ArrayList;

/**
 *
 * @author angel
 */

    
public class Jugador implements Comparable<Jugador>{
    
    protected static final int CASASMAX = 4;
    protected static final int CASASPORHOTEL = 4;
    protected boolean encarcelado;
    protected static final int HOTELESMAX = 4;
    private String nombre;
    private int numCasillaActual;
    protected static final float PASOPORSALIDA = 100;
    protected static final float PRECIOLIBERTAD = 200;
    private boolean puedeComprar;
    private float saldo;
    private static final float SALDOINICIAL = 7500;
    private ArrayList<TituloPropiedad> propiedades;
    private Sorpresa salvoconducto;
    
    @Override
    public int compareTo(Jugador otro) {
        if (saldo > otro.saldo) {
            return -1;
        }
        if (saldo < otro.saldo) {
            return 1;
        }
        return 0;

    }
    
    Jugador(String nombre){
        
        encarcelado = false;
        this.nombre = nombre;
        numCasillaActual = 0;
        puedeComprar = true;
        saldo = SALDOINICIAL;
        propiedades = new ArrayList<TituloPropiedad>();
        salvoconducto = null;
        
    }
    
    protected Jugador (Jugador jugador){
        encarcelado = jugador.encarcelado;
        nombre = jugador.nombre;
        numCasillaActual = jugador.numCasillaActual;
        puedeComprar = jugador.puedeComprar;
        saldo = jugador.saldo;
        salvoconducto = jugador.salvoconducto;
        /*
        propiedades = new ArrayList<>();
        
        if (!jugador.propiedades.isEmpty())
            for (int i = 0; i < jugador.propiedades.size(); i++)
                propiedades.set(i, jugador.propiedades.get(i));
        else
            propiedades = null;
        */
        propiedades= jugador.propiedades;
    }
    
    protected boolean debeSerEncarcelado(){
        
        if (!encarcelado){
            if (salvoconducto == null)
                return true;
            else{
                perderSalvoconducto();
                Diario.instance.ocurreEvento("El jugador " + nombre + " se libra de la carcel.");
                return false;
                
            }
                
        }
        else
            return false;
        
    }
    
    boolean encarcelar(int numCasillaCarcel){
        if (debeSerEncarcelado()){
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            Diario.instance.ocurreEvento("El jugador " + nombre + " ha sido trasladado a la carcel.");
        }
        else
            Diario.instance.ocurreEvento("El jugador " + nombre + " no ha sido trasladado a la carcel.");
        
        return encarcelado;
    }
    
    boolean obtenerSalvoconducto(Sorpresa sorpresa){
        
        if (!encarcelado){
            salvoconducto = sorpresa;
            return true;
        }
        else
            return false;
    }
    
    private void perderSalvoconducto(){
        salvoconducto.usada();
        salvoconducto = null;
    }
    
    boolean tieneSalvoconducto(){
        return salvoconducto != null;
    }
    
    boolean puedeComprarCasilla(){
        puedeComprar = !encarcelado;
        return puedeComprar;
    }
    
    boolean paga(float cantidad){
        int j = 0;
        boolean jaja = modificarSaldo(cantidad*(-1));
        int x = 0;
        return jaja;
    }
    
    boolean pagaImpuesto(float cantidad){
        if (encarcelado)
            return false;
        else
            return paga(cantidad);
    }
    
    boolean pagaAlquiler(float cantidad){
        if (encarcelado)
            return false;
        else
            return paga(cantidad);
    }
    
    boolean recibe(float cantidad){
        if (encarcelado)
            return false;
        else{
            int x = 0;
            boolean m = modificarSaldo(cantidad);
            int j = 0;
            return m;
            //return modificarSaldo(cantidad);
            
        }
    }
    
    boolean modificarSaldo(float cantidad){
        saldo += cantidad;
        Diario.instance.ocurreEvento("El saldo del jugador " + nombre + " ha sido modificado a " + saldo);
        return true;
    }
    
    boolean moverACasilla(int numCasilla){
        if (!encarcelado){
            numCasillaActual = numCasilla;
            puedeComprar = false;
            Diario.instance.ocurreEvento("El jugador " + nombre + " ha sido movido a la casilla " + numCasilla);
           return true;
        }
        else{
            Diario.instance.ocurreEvento("El jugador " + nombre + " no ha sido movido porque esta encarcelado");
            return false;
        }
    }
    
    private boolean puedoGastar(float precio){
        if (!encarcelado)
            return saldo >= precio;
        else
            return false;
    }
    
    boolean vender(int ip){
        
        if (!encarcelado){
            if(existeLaPropiedad(ip)){
                boolean vender = propiedades.get(ip).vender(this);
                if(vender){
                    Diario.instance.ocurreEvento("El jugador " + nombre + " ha vendido la propiedad " + propiedades.get(ip).getNombre());
                    propiedades.remove(ip);
                    return true;
                }
                
            }
       
        
        }
            
        Diario.instance.ocurreEvento("El jugador " + nombre + " no ha podido vender la propiedad (porque estaba encarcelada o no es suya) " + propiedades.get(ip).getNombre());       
        return false;
    }
    boolean tieneAlgoQueGestionar(){
        return propiedades != null;
    }
    
    boolean puedeSalirCarcelPagando(){
        return saldo >= PRECIOLIBERTAD;
    }
    
    boolean salirCarcelPagando(){
        if (encarcelado && puedeSalirCarcelPagando()){
            paga(PRECIOLIBERTAD);
            encarcelado = false;
            Diario.instance.ocurreEvento("El jugador " + nombre + " sale de la carcel pagando\n");
            return true;
        }

        Diario.instance.ocurreEvento("El jugador " + nombre + " no ha podido salir de la carcel pagando (no tiene money) \n");
        return false;
        
    }
    
    boolean salirCarcelTirando(){
        if(Dado.instance.salgoDeLaCarcel()){
            encarcelado = false;
            Diario.instance.ocurreEvento("El jugador " + nombre + " sale de la carcel.");
            return true;
        }   
        
        Diario.instance.ocurreEvento("El jugador " + nombre + " ha tirado pero no sale de la carcel.");
        return false;
    }
    
    
    boolean pasaPorSalida(){
        int x = 0;
        modificarSaldo(PASOPORSALIDA);
        Diario.instance.ocurreEvento("El jugador " + nombre + " pasa por salida y cobra 100.");
        return true;
    }
    
    int cantidadCasasHoteles(){
        
        int suma = 0;
        
        for (int i = 0; i < propiedades.size(); i++)
            suma += propiedades.get(i).cantidadCasasHoteles();
        
        return suma;
    }
    
    boolean enBancarrota(){
        return saldo < 0;
    }
    
    private boolean existeLaPropiedad(int ip){
        return ip >= 0 && ip < propiedades.size();
    }
    
    private int getCasasMax(){
        return CASASMAX;
    }
    
    int getCasasPorHotel(){
        return CASASPORHOTEL;
    }
    
    private int getHotelesMax(){
        return HOTELESMAX;
    }
    
    protected String getNombre(){
        return nombre;
    }
    
    int getNumCasillaActual(){
        return numCasillaActual;
    }
    
    private float getPrecioLibertad(){
        return PRECIOLIBERTAD;
    }
    
    private float getPremioPasoPorSalida(){
        return PASOPORSALIDA;
    }
    
    public ArrayList<TituloPropiedad> getPropiedades(){
        return propiedades;
    }
    
    boolean getPuedeComprar(){
        return puedeComprar;
    }
    
    protected float getSaldo(){
        return saldo;
    }
    
    public boolean isEncarcelado(){
        return encarcelado;
    }
    
    private boolean puedoEdificarCasa(TituloPropiedad propiedad){
        
        float precio = propiedad.getPrecioEdificar();
        
        return propiedad.esEsteElPropietario(this) && propiedad.getNumCasas() < CASASPORHOTEL && this.puedoGastar(precio);
        
    }
    
    private boolean puedoEdificarHotel(TituloPropiedad propiedad){
        
        float precio = propiedad.getPrecioEdificar();
        
        return propiedad.getNumCasas() ==  CASASPORHOTEL && propiedad.getNumHoteles() < CASASPORHOTEL && this.puedoGastar(precio);
    }
    
    boolean cancelarHipoteca(int ip){
    
        boolean result=false;
        
        if(encarcelado){
            
            return result;
            
        }
        if(existeLaPropiedad(ip)){
            
            TituloPropiedad propiedad = propiedades.get(ip);  
            float cantidad = propiedad.getImporteCancelarHipoteca();
            boolean puedoGastar=puedoGastar(cantidad);
            if(puedoGastar){
                
                result=propiedad.cancelarHipoteca(this);
            
                
                if(result){
                    
                Diario.getInstance().ocurreEvento("El jugador "+nombre+ " cancela la hipoteca de la propiedad " +propiedades.get(ip).getNombre());
                    
                }
                else{
                    Diario.getInstance().ocurreEvento("El jugador "+nombre+ " no puede cancelar la hipoteca de la propiedad " +propiedades.get(ip).getNombre());
                }
                
                
            }else
                Diario.getInstance().ocurreEvento("El jugador "+nombre+ " no puede cancelar la hipoteca de la propiedad " +propiedades.get(ip).getNombre());
            
        }
        
        
        return result;
    
    }
    
    
       boolean comprar(TituloPropiedad titulo){
           
           boolean result = false;

        if (encarcelado)
            return result;


        if (puedeComprar){

            float precio = titulo.getPrecioCompra();

            if (puedoGastar(precio)){
                
                result = titulo.comprar(this);

                if (result){
                    propiedades.add(titulo);
                    Diario.instance.ocurreEvento("El jugador "+nombre+ " compra " + titulo.toString());
                }
                
                puedeComprar = false;
                                
            }
            
            

       }

        return result;
           
           
        }
    
     
    
    Boolean construirHotel (int ip){
    
    boolean result = false;
        boolean puedoEdificarHotel;
        float precio;
        int casasPorHotel;
        
        if(encarcelado)
            return result;
        
        if(existeLaPropiedad(ip)){
            
            
            TituloPropiedad propiedad;
            propiedad = propiedades.get(ip);
            
            
            puedoEdificarHotel=puedoEdificarHotel(propiedad);
            
            if(puedoEdificarHotel){
                
                result = propiedad.construirHotel(this);
                
                
                casasPorHotel=getCasasPorHotel();
                
                propiedad.derruirCasas(casasPorHotel, this);
                Diario.getInstance().ocurreEvento("El jugador "+nombre+" construye hotel en la propiedad "+ip);
            }
            else
                Diario.getInstance().ocurreEvento("El jugador "+nombre+" no construye hotel en la propiedad "+ip + " porque no tiene 4 casas construidas o no tiene dinero.\n");
        }
        
        return result;
    
    }
    
    
    Boolean construirCasa(int ip){
        
        boolean result =false, puedoEdificarCasa =false;
        
        if(encarcelado){
            return result;
        }else{
            
            boolean existe = existeLaPropiedad(ip);
        
            if(existe){
                
                TituloPropiedad propiedad=propiedades.get(ip);
                puedoEdificarCasa=puedoEdificarCasa(propiedad);
                float precio=propiedad.getPrecioEdificar();
                if(puedoGastar(precio) && propiedad.getNumCasas()<getCasasMax()){
                    
                    puedoEdificarCasa=true;
                    
                }
                if(puedoEdificarCasa){
                    
                    result=propiedad.construirCasa(this);
                    
                    if(result){
                        
                        Diario.getInstance().ocurreEvento("El jugador " +nombre+ " construye casa en la propiedad "+ propiedades.get(ip));
                        
                    }
                    else{
                        Diario.getInstance().ocurreEvento("El jugador " +nombre+ " no puede construir casa en la propiedad (porque ya tiene 4 o no tiene money) "+ propiedades.get(ip) + "\n");
                    }
                }
                else
                    Diario.getInstance().ocurreEvento("El jugador " +nombre+ " no puede construir casa en la propiedad (porque ya tiene 4 o no tiene money) "+ propiedades.get(ip) + "\n");
            }
        }
        
        return result;
    }
    
    protected boolean hipotecar(int ip){
        
        boolean result=false;
        if(encarcelado){
            
            return result;
            
        }
        
        if(existeLaPropiedad(ip)){
            result = propiedades.get(ip).hipotecar(this);
                     
        }
        if(result){
            
            Diario.getInstance().ocurreEvento("El jugador " +nombre+ " hipoteca la propiedad "+ propiedades.get(ip).getNombre());
        }
        else
            Diario.getInstance().ocurreEvento("El jugador " +nombre+ " no puede hipotecar la propiedad "+ propiedades.get(ip).getNombre());
        
        return result;
    }
    
    
    @Override
    public String toString(){
        
        return "Jugador de nombre " + nombre + " situado en la casilla " + numCasillaActual + " con saldo " + saldo;
    }
    
    
}
